### PR TITLE
Auto-grown sparse table

### LINKED ISSUE #9841: Need a new class to represent some certain interface of Table used in lookup remote table
For the feature **lookup remote table**, we use the variable type `SelectedRows` to represent a `Table` currently. But sometimes the code is not certainly enough. Maybe we need a new class named `Table`(or othes) that inherit `SelectedRows` and implement some certain interface, the pseudo code as following:
```cpp
class Table : public framework::SelectedRows {
bool HasKey(int64_t key)
bool Resize(framework::ddim shape); // resize value and rows size
Tensor Get(int64_t key)
bool Set(int64_t key, Tensor value);
...
}
```

### PR DESCRIPTION
**Pull request by Yancey1989**
Fixed #9841

### CODE REVIEW

File: paddle/fluid/framework/selected_rows.cc [hunk 1]
```diff
@@ -69,5 +116,66 @@ void DeserializeFromStream(std::istream& is, SelectedRows* selected_rows,
   TensorFromStream(is, selected_rows->mutable_value(), dev_ctx);
 }
 
+bool SelectedRows::HasKey(int64_t key) const {
```
Comment (typhoonzero) [inline paddle/fluid/framework/selected_rows.cc]: Maybe add some comments here to explain why we need these APIs for `SelectedRows` data structure.

File: paddle/fluid/framework/selected_rows.cc [hunk 2]
```diff
@@ -69,5 +116,66 @@ void DeserializeFromStream(std::istream& is, SelectedRows* selected_rows,
...
+  return std::find(rows_.begin(), rows_.end(), key) == rows_.end() ? false
+                                                                   : true;
+}
+
+bool SelectedRows::Get(int64_t key, framework::Tensor* value,
+                       int64_t row) const {
```
Comment (typhoonzero) [inline paddle/fluid/framework/selected_rows.cc]: row is the same as key?

File: paddle/fluid/framework/selected_rows.cc (hunk 1 above)
Comment (Yancey1989) [reply paddle/fluid/framework/selected_rows.cc]: Done.

File: paddle/fluid/framework/selected_rows.cc (hunk 2 above)
Comment (Yancey1989) [reply paddle/fluid/framework/selected_rows.cc]: No, it's the offset of the given value pointer, maybe we can rename it to `offset`...

Comment (Yancey1989) [reply paddle/fluid/framework/selected_rows.cc]: Done.

File: paddle/fluid/framework/selected_rows.cc [hunk 3]
```diff
@@ -17,6 +17,53 @@ limitations under the License. */
...
+
+  template <typename T>
+  void operator()() const {
+    framework::Tensor cpu_tensor;
+    platform::CPUPlace cpu;
+    T* ptr = cpu_tensor.mutable_data<T>(dims_, cpu);
+    const T* old_ptr =
+        tensor_->memory_size() == 0 ? nullptr : tensor_->data<T>();
+    if (old_ptr != nullptr) {
+      std::copy(old_ptr, old_ptr + tensor_->numel(), ptr);
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply paddle/fluid/framework/selected_rows.cc]: Done.

**Pull request merged**